internal/buffer/save.go

```diff
@@ -184,59 +263,13 @@ func (b *Buffer) saveToFile(filename string, withSudo bool, autoSave bool) error
 		}
 	}
 
-	var fileSize int
-
-	enc, err := htmlindex.Get(b.Settings["encoding"].(string))
-	if err != nil {
-		return err
-	}
-
-	fwriter := func(file io.Writer) (e error) {
-		if len(b.lines) == 0 {
-			return
-		}
-
-		// end of line
-		var eol []byte
-		if b.Endings == FFDos {
-			eol = []byte{'\r', '\n'}
-		} else {
-			eol = []byte{'\n'}
-		}
-
-		// write lines
-		if fileSize, e = file.Write(b.lines[0].data); e != nil {
-			return
-		}
-
-		for _, l := range b.lines[1:] {
-			if _, e = file.Write(eol); e != nil {
-				return
-			}
-			if _, e = file.Write(l.data); e != nil {
-				return
-			}
-			fileSize += len(eol) + len(l.data)
-		}
-		return
-	}
-
-	if err = overwriteFile(absFilename, enc, fwriter, withSudo); err != nil {
+	w := BufferWriter{b, withSudo}
+	if err = util.SafeWrite(absFilename, w); err != nil {
```

Comment:
So I've been testing this, and it is still not really safe or useful. I try to save a file on an almost full tmpfs filesystem, micro says it fails to save it due to no space left on device, the file becomes corrupted (written imcompletely), but before that micro indeed successfully writes it to the backup file in `~/.config/micro/backups`, so I should be able at least to recover the file next time I open it in micro? But no, I can't, since after I close the buffer, the backup file in `~/.config/micro/backups` disappears (probably because we always call `RemoveBackup()` in `b.Fini()`).